GA 2.0 release notes (#605)
dotnet/SqlClient: release-notes/2.0/2.0.0.md

### Changes
- Raise warning message when insecure TLS protocols are in use [#591](https://github.com/dotnet/SqlClient/pull/591)

### Breaking Changes
- Modified enclave provider interface `SqlColumnEncryptionEnclaveProvider` to be internal [#602](https://github.com/dotnet/SqlClient/pull/602) - _Since Secure enclaves is a relatively new feature, this change is not likely to impact customer applications_.
- Updated `SqlClientMetaDataCollectionNames` exposed constants by removing non-existing constants and adding new to the metadata collection [#580](https://github.com/dotnet/SqlClient/pull/580)

## New features in v2.0

New features over the 1.1 release of Microsoft.Data.SqlClient:
- [Additional Active Directory authentication modes](#additional-active-directory-authentication-modes)
- [EventSource tracing support](#eventSource-tracing-support)
- [Enable Managed networking on Windows](#enable-managed-networking-on-windows)
- [Enable decimal truncation behavior conditionally](#enable-decimal-truncation-behavior-conditionally)
- [New connection string property synonyms](#new-connection-string-property-synonyms)

> 

Reviewer: SqlBulkCopy fixes??

### Additional Active Directory authentication modes
This release brings parity of active directory authentication modes supported for .NET Framework and .NET Core applications. With 2.0 stable release, below authentication modes are supported for **Microsoft.Data.SqlClient**:

|Authentication mechanism | .NET Framework 4.6+ | .NET Core 2.1+ | .NET Standard 2.0+|
|------|--------|--------|--------|
|Active Directory Password | Yes | Yes | Yes |
|Active Directory Integrated | Yes | **Yes**<sup>1</sup>  | **Yes**<sup>1</sup>  |
|Active Directory Interactive | Yes | **Yes**<sup>1</sup>  | **Yes**<sup>1</sup>  |
|**Active Directory Service Principal**<sup>1</sup> | **Yes**<sup>1</sup>  | **Yes**<sup>1</sup>  | **Yes**<sup>1</sup>  |

_<sup>1</sup> New authentication mode starting with Microsoft.Data.SqlClient v2.0_

#### Active Directory Service Principal

> This authentication mode uses Active Directory Service Principal to connect to a SQL Database using the client ID and secret of a service principal identity. Service principal authentication involves setting up an App registration with a secret, granting permissions to the App in the SQL DB instance, then connecting with connection string as under: 

Reviewer: Mention "Azure" ?

- Added support for `Active Directory Service Principal` authentication mode for .NET Framework [#560](https://github.com/dotnet/SqlClient/pull/560)

### Fixed
- Fixed `SqlSequentialStream` multipacket read stalling issue in .NET Core [#603](https://github.com/dotnet/SqlClient/pull/603)
- Fixed code page issue for Kazakh collation in SQL Server [#584](https://github.com/dotnet/SqlClient/pull/584)
- Fixed stalled application issues when end of stream is reached [#577](https://github.com/dotnet/SqlClient/pull/577)
- Fixed driver behavior to not throw exception for invalid configuration file [#573](https://github.com/dotnet/SqlClient/pull/573)
- Fixed Object null reference issue when failover partner is set [#588](https://github.com/dotnet/SqlClient/pull/588)
- Fixed `applicationintent` connection string property issue [#585](https://github.com/dotnet/SqlClient/pull/585)

### Changes
- Raise warning message when insecure TLS protocols are in use [#591](https://github.com/dotnet/SqlClient/pull/591)

### Breaking Changes

> - Modified enclave provider interface `SqlColumnEncryptionEnclaveProvider` to be internal [#602](https://github.com/dotnet/SqlClient/pull/602) - _Since Secure enclaves is a relatively new feature, this change is not likely to impact customer applications_.

Reviewer 2: ```suggestion
- Modified enclave provider interface `SqlColumnEncryptionEnclaveProvider` to be internal [#602](https://github.com/dotnet/SqlClient/pull/602) - _This change is not likely to impact customer applications since secure enclaves is a relatively new feature and they would have had to implement their own enclave provider, which is not a trivial task_.
```

- Updated `SqlClientMetaDataCollectionNames` exposed constants by removing non-existing constants and adding new to the metadata collection [#580](https://github.com/dotnet/SqlClient/pull/580)

## New features in v2.0

New features over the 1.1 release of Microsoft.Data.SqlClient:
- [Additional Active Directory authentication modes](#additional-active-directory-authentication-modes)
- [EventSource tracing support](#eventSource-tracing-support)
- [SNI dependency changes](#sni-dependency-changes)
- [Enable Managed networking on Windows](#enable-managed-networking-on-windows)
- [Enable decimal truncation behavior conditionally](#enable-decimal-truncation-behavior-conditionally)
- [New connection string property synonyms](#new-connection-string-property-synonyms)
- [SqlBulkCopy RowsCopied property](#sqlbulkcopy.rowscopied-property)
- [Connection Open Overrides](#connection-open-overrides)
- [Username support for Active Directory Interactive mode](#username-support-for-active-directory-interactive-mode)
- [Order Hints for SqlBulkCopy](#order-hints-for-sqlbulkcopy)

### Additional Active Directory authentication modes

> This release brings parity of active directory authentication modes supported for .NET Framework and .NET Core applications. With 2.0 stable release, below authentication modes are supported for **Microsoft.Data.SqlClient**:

Reviewer 2: ```suggestion
This release brings parity of active directory authentication modes supported for .NET Framework and .NET Core applications. With the 2.0 stable release, the following authentication modes are supported for **Microsoft.Data.SqlClient**:
```

> This release brings parity of active directory authentication modes supported for .NET Framework and .NET Core applications. With the 2.0 stable release, the following authentication modes are supported for **Microsoft.Data.SqlClient**:

Reviewer 3: ```suggestion
This release brings parity of Active Directory authentication modes supported for .NET Framework and .NET Core applications. With the 2.0 stable release, the following authentication modes are supported for **Microsoft.Data.SqlClient**:
```

|Authentication mechanism | .NET Framework 4.6+ | .NET Core 2.1+ | .NET Standard 2.0+|
|------|--------|--------|--------|
|Active Directory Password | Yes | Yes | Yes |
|Active Directory Integrated | Yes | **Yes**<sup>1</sup>  | **Yes**<sup>1</sup>  |
|Active Directory Interactive | Yes | **Yes**<sup>1</sup>  | **Yes**<sup>1</sup>  |
|**Active Directory Service Principal**<sup>1</sup> | **Yes**<sup>1</sup>  | **Yes**<sup>1</sup>  | **Yes**<sup>1</sup>  |

_<sup>1</sup> New authentication mode starting with Microsoft.Data.SqlClient v2.0_

#### Active Directory Service Principal

> This authentication mode uses Active Directory Service Principal to connect to a Azure Database using the client ID and secret of a service principal identity. Service principal authentication involves setting up an App registration with a secret, granting permissions to the App in the SQL DB instance, then connecting with connection string as under: 

Reviewer 2: ```suggestion
This authentication mode uses Active Directory Service Principal to connect to an Azure Database using the client ID and secret of a service principal identity. Service principal authentication involves setting up an App registration with a secret, granting permissions to the App in the SQL DB instance, then connecting with a connection string like the following: 
```

| ExecutionTrace | 1 | Turns on capturing Start/Stop events before and after command execution. |
| Trace | 2 | Turns on capturing basic application flow trace events. |
| Scope | 4 | Turns on capturing enter and exit events |
| NotificationTrace | 8 | Turns on capturing `SqlNotification` trace events |
| NotificationScope | 16 | Turns on capturing `SqlNotification` scope enter and exit events |
| PoolerTrace | 32 | Turns on capturing connection pooling flow trace events. |
| PoolerScope | 64 | Turns on capturing connection pooling scope trace events. |
| AdvancedTrace | 128 | Turns on capturing advanced flow trace events. |
| AdvancedTraceBin  | 256 | Turns on capturing advanced flow trace events with additional information. |
| CorrelationTrace | 512 | Turns on capturing correlation flow trace events. |
| StateDump | 1024 | Turns on capturing full state dump of `SqlConnection` |
| SNITrace | 2048 | Turns on capturing flow trace events from Managed Networking implementation (only applicable in .NET Core) |
| SNIScope | 4096 | Turns on capturing scope events from Managed Networking implementation (only applicable in .NET Core) |
|||

### Enable managed networking on Windows

> This release introduces a new AppContext switch "Switch.Microsoft.Data.SqlClient.UseManagedNetworkingOnWindows" that enables the use of Managed SNI on Windows for testing and debugging purposes. This switch will toggle the driver's behavior to use Managed SNI in .NET Core 2.1+ and .NET Standard 2.0+ projects on Windows.

Reviewer 2: ```suggestion
This release introduces a new AppContext switch, "Switch.Microsoft.Data.SqlClient.UseManagedNetworkingOnWindows", that enables the use of Managed SNI on Windows for testing and debugging purposes. This switch will toggle the driver's behavior to use Managed SNI in .NET Core 2.1+ and .NET Standard 2.0+ projects on Windows. Using the managed SNI implementation eliminates the dependency on the native Microsoft.Data.SqlClient.SNI binaries for a fully managed stack.
```
